tenpy/algorithms/tebd.py

```diff
@@ -591,6 +591,264 @@ def __init__(self, psi, model, options):
         TEBDEngine.__init__(self, psi, model, options)
 
 
+class QRBasedTEBDEngine(TEBDEngine):
+    r"""Version of :class:`~tenpy.algorithms.tebd.TEBDEngine` that relies on QR decompositions
+    for the truncation of the evolved local wavefunction.
+    As introduced in :arxiv:`2212.09782`.
+
+    .. todo ::
+        To use `use_eig_based_svd == True`, which makes sense on GPU only, we need to implement
+        the `_eig_based_svd` for "non-square" matrices.
+        This means that :math:`M^{\dagger} M` and :math:`M M^{\dagger}` dont have the same size,
+        and we need to disregard those eigenvectors of the larger one, that have eigenvalue zero,
+        since we dont have corresponding eigenvalues of the smaller one.
+
+    .. todo ::
+        Can we implement some simple diagnostics that issue a warning if the CBE expansion
+        was too small?
+        Could potentially look into the tail of the schmidt spectrum.
```

Comment:
I don't think that's necessary. Are you afraid that the expansion factor was too small? It can grow exponentially with the number of steps, so it should quickly be able to catch up to the needed bond dimension.

Reply:
I am afraid about what happens when we start from a low bond dimension MPS, e.g. a bond dimension. If we have a large on-site Hilbertspace, which is exactly where QR TEBD may be advantageous, we have $d^2$ schmidt values in the exact evolved state but we keep only 2 singular values, if `expand <= 1` as it probably will be. I worry that the schmidt values will not have sufficiently decayed in this case, since the horizontal axis of the entanglement spectrum is quite coarse.

In any case, such problems would manifest in large truncation errors, so they would be detected. So I will remove this.

Reply:
If you are really worried by this, you can increase the `max(1, ...)` to `max(10, ...)` in line 738 of tebd.py; I think the rest should go through. This will help to much faster increase initially tiny when needed, but the downside is that ithe total expanded size (before truncation) could become significantly larger than `expand * old_chi` if you have many tiny blocks. We could also make it a full parameter (rather than a hard-coded magic parameter).

Reply:
I will make it a paramter 👍